fluent.runtime/fluent/runtime/__init__.py

```diff
@@ -44,22 +47,38 @@ def add_messages(self, source):
             if isinstance(item, (Message, Term)):
                 full_id = ast_to_id(item)
                 if full_id not in self._messages_and_terms:
-                    # We add attributes to the store to enable faster looker
-                    # later, and more direct code in some instances.
-                    add_message_and_attrs_to_store(self._messages_and_terms, full_id, item)
+                    self._messages_and_terms[full_id] = item
 
     def has_message(self, message_id):
         if message_id.startswith(TERM_SIGIL) or ATTRIBUTE_SEPARATOR in message_id:
             return False
         return message_id in self._messages_and_terms
 
+    def lookup(self, full_id):
+        if full_id not in self._compiled:
+            entry_id = full_id.split(ATTRIBUTE_SEPARATOR, 1)[0]
+            entry = self._messages_and_terms[entry_id]
+            compiled = self._compiler(entry)
+            if compiled.value is not None:
+                self._compiled[entry_id] = compiled.value
+            for attr in compiled.attributes:
+                self._compiled[ATTRIBUTE_SEPARATOR.join([entry_id, attr.id.name])] = attr.value
+        return self._compiled[full_id]
+
     def format(self, message_id, args=None):
         if message_id.startswith(TERM_SIGIL):
             raise LookupError(message_id)
-        message = self._messages_and_terms[message_id]
-        if args is None:
-            args = {}
-        return resolve(self, message, args)
+        fluent_args = {}
+        if args is not None:
+            for argname, argvalue in args.items():
+                fluent_args[argname] = native_to_fluent(argvalue)
```

Comment:
nit: this looks like it could be written more succinctly using a dict comprehension

    fluent_args = ({} if args is None else
                   {argname: native_to_fluent(argvalue) for argname, argvalue in args.items()})

Reply:
I've done half of this, as I found the inline `if` to be too hard to read. I made a top-level `if`, and then the dict comprehension or an empty dict.